### PR TITLE
Feat/twitter follow profile

### PR DESCRIPTION
**Pull request by rafagomes**
# Twitter follow Task

## Type of PR

- [] - feature

## Description
A feature that allows the user to verify if he followed the credential task twitter.

### CODE REVIEW

File: apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    getValues,
+    setError,
+    formState: { errors },
+  } = useFormContext<CreateGateTypes>();
+
+  useEffect(() => {
+    setValue(`tasks.data.${taskId}.title`, 'Untitled Task');
+  }, [setValue, taskId]);
+
+  const getTwitterData = async () => {
```
Comment (kbooz) [inline apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx:61]: Use react-query for handling the data and the loading state

File: apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        >
+          <CircleWithNumber
+            number={taskId + 1}
+            sx={(theme) => ({
+              mr: theme.spacing(3.75),
+              [theme.breakpoints.down('sm')]: { mr: theme.spacing(2.5) },
+            })}
+          />
+          <Stack>
+            <Typography variant="subtitle2">Follow Profile</Typography>
```
Comment (kbooz) [inline apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx:139]: put content on locale

File: apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+  readOnly,
+  isLoading,
+  isAdmin,
+}) => {
+  const [twitterKeys, setTwitterKeys] = useState(null);
+  const { gqlAuthMethods } = useAuth();
+  const formattedDate = new Date(updatedAt.toLocaleString()).toLocaleString();
+  const [twitterData, setTwitterData] = useState(null);
+
+  const getTwitterData = async () => {
```
Comment (kbooz) [inline apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx:28]: use react-query to handle data fetch

File: apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    try {
+      const username = data.username;
+      const response = await gqlAuthMethods.twitter_data({
+        userName: username,
+      });
+
+      return setTwitterData(response.get_twitter_user_data);
+    } catch (error) {
+      console.log(error);
```
Comment (kbooz) [inline apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx:37]: console log

File: apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    }
+  };
+
+  const connectTwitter = async () => {
```
Comment (kbooz) [inline apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx:41]: use react-query to handle data fetch

File: apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    } else if (value < 1000000) {
+      return `${(value / 1000).toFixed(1)}K`;
+    } else if (value < 1000000000) {
+      return `${(value / 1000000).toFixed(1)}M`;
+    }
+  };
+
+  useEffect(() => {
+    if (typeof window !== 'undefined') {
+      const twitter = JSON.parse(window.localStorage.getItem('twitter'));
```
Comment (kbooz) [inline apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx:95]: `react-use` has a way to handle localStorage
https://github.com/streamich/react-use/blob/master/docs/useLocalStorage.md

File: apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+                  sx={{
+                    color: '#0F1419',
+                    size: '1rem',
+                    fontWeight: 'bold',
+                    fontFamily: 'sans-serif',
+                  }}
+                >
+                  {numberFormat(twitterData?.public_metrics.following_count)}{' '}
+                  <Typography component="span" sx={{ color: '#5B7083' }}>
+                    Following
```
Comment (kbooz) [inline apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx:185]: content on locale

File: apps/website/pages/following.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import { useEffect } from 'react';
+
+import { useLocation } from 'react-use';
+
+export default function RedirectPage() {
```
Comment (kbooz) [inline apps/website/pages/following.tsx:5]: wound't it be better if this page url would be smth like `/oauth/twitter`, for managing other callbacks

File: apps/website/pages/following.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  const search = useLocation().search;
+  const oauthToken = new URLSearchParams(search).get('oauth_token');
+  const oauthVerifier = new URLSearchParams(search).get('oauth_verifier');
+
+  useEffect(() => {
```
Comment (kbooz) [inline apps/website/pages/following.tsx:10]: use react-query and useLocalStorage to handle this

**Review by kbooz — CHANGES_REQUESTED** [review]

File: apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx (hunk 2 above)
Comment (rafagomes) [reply apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx:139]: will do on a future PR

File: apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx (hunk 7 above)
Comment (rafagomes) [reply apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx:182]: on a future PR

File: apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+    getValues,
+    setError,
+    formState: { errors },
+  } = useFormContext<CreateGateTypes>();
+
+  useEffect(() => {
+    setValue(`tasks.data.${taskId}.title`, 'Untitled Task');
+  }, [setValue, taskId]);
+
+  const getTwitterData = useQuery(
```
Comment (kbooz) [inline apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx:61]: if this is being called later, use a useMutation instead

File: apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+        return setError(`tasks.data.${taskId}.task_data.username`, {
+          message: 'User not found',
+        });
+      }
+    },
+    { enabled: false }
+  );
+
+  const delayedQuery = useCallback(
+    debounce(() => getTwitterData.refetch(), 500),
```
Comment (kbooz) [inline apps/website/components/molecules/add-task/twitter-follow-profile/twitter-follow-profile.tsx:85]: change getTwitterData to a `useMutation` and call `.mutate()` instead

File: apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+      } catch (error) {
+        console.log(error);
+      }
+    },
+    {
+      enabled: false,
+    }
+  );
+
+  const checkTwitterFollow = useQuery(
```
Comment (kbooz) [inline apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx:64]: change this to a `useMutation`

File: apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+      const response = await gqlAuthMethods.twitter_data({
+        userName: username,
+      });
+      return response.get_twitter_user_data;
+    } catch (error) {
+      throw Error(error);
+    }
+  });
+
+  const connectTwitter = useQuery(
```
Comment (kbooz) [inline apps/website/components/organisms/gates/view/tasks/content/twitter-follow.tsx:44]: change this to a `useMutation`

**Review by kbooz — CHANGES_REQUESTED** [review]

**Pull request merged**